From 662800081ef7b647a742f32eecf7317fa276d20b Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Wed, 29 May 2024 10:42:11 +0200
Subject: [PATCH] drm/stm: ltdc: flush remaining vblank event

When closing the crtc, check the crtc status and
send an event if necessary.

Change-Id: I791f8e9b8b05084f50f184ecf50067adbf90eb61
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/385854
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/stm/ltdc.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/gpu/drm/stm/ltdc.c b/drivers/gpu/drm/stm/ltdc.c
index 5ed71dfcb2ae..b3a0dca98340 100644
--- a/drivers/gpu/drm/stm/ltdc.c
+++ b/drivers/gpu/drm/stm/ltdc.c
@@ -1114,6 +1114,14 @@ static void ltdc_crtc_atomic_disable(struct drm_crtc *crtc,
 	ldev->fifo_warn = 0;
 	ldev->fifo_rot = 0;
 	mutex_unlock(&ldev->err_lock);
+
+	/* Flush remaining vblank event*/
+	if (crtc->state->event && !crtc->state->active) {
+		spin_lock_irq(&crtc->dev->event_lock);
+		drm_crtc_send_vblank_event(crtc, crtc->state->event);
+		spin_unlock_irq(&crtc->dev->event_lock);
+		crtc->state->event = NULL;
+	}
 }
 
 static void ltdc_crtc_atomic_flush(struct drm_crtc *crtc,
-- 
2.39.5

